server/gcpcredential/validate_test.go

```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		t.Fatalf("Error generating test key: %v", err)
+	}
+
+	jwk := JWK{
+		Alg: "RS256",
+		Kid: keyID,
+		N:   base64.RawURLEncoding.EncodeToString(key.N.Bytes()),
+		E:   base64.RawURLEncoding.EncodeToString(big.NewInt(int64(key.E)).Bytes()),
+	}
+
+	return key, jwk
+}
+
+// Generates and returns a JWT token with the provided claims, audience, and signed by the provided key.
+func testGCPCredential(t *testing.T, claims *emailClaims, aud, keyID string, signer *rsa.PrivateKey) string {
+	t.Helper()
+
+	now := time.Now().Unix()
+
+	jwtClaims := jwt.MapClaims{
+		"aud":            aud,
+		"iss":            "accounts.google.com",
+		"exp":            now + 60,
+		"email":          claims.Email,
+		"email_verified": claims.EmailVerified,
+	}
+
+	token := jwt.NewWithClaims(jwt.SigningMethodRS256, jwtClaims)
+	token.Header = map[string]any{
+		"kid": keyID,
+		"alg": "RS256",
+	}
+	tokenString, err := token.SignedString(signer)
+	if err != nil {
+		t.Fatalf("Error generating token for %v: %v", claims, err)
+	}
+
+	return tokenString
+}
+
+type jwkFetcher struct {
+	jwkFunc func(req *http.Request) *http.Response
+}
+
+func (t *jwkFetcher) RoundTrip(req *http.Request) (*http.Response, error) {
+	return t.jwkFunc(req), nil
+}
+
+func testJWKFetcher(t *testing.T, jwks *JWKS) *jwkFetcher {
+	t.Helper()
+	return &jwkFetcher{
+		jwkFunc: func(req *http.Request) *http.Response {
+			respBytes, err := json.Marshal(jwks)
+			if err != nil {
+				t.Fatalf("Unable to marshal server response: %v", err)
+			}
+
+			return &http.Response{
+				StatusCode: http.StatusOK,
+				Header:     make(http.Header),
+				Body:       ioutil.NopCloser(bytes.NewBuffer(respBytes)),
+			}
+		},
+	}
+}
+
+func TestValidation(t *testing.T) {
+	signerA, jwkA := testRSASigner(t, testKeyID+"A")
+	signerB, jwkB := testRSASigner(t, testKeyID+"B")
+	jwks := &JWKS{[]JWK{jwkA, jwkB}}
+
+	expectedEmails := []string{"tokenA@test.com", "tokenB@test.com"}
+	testTokens := []string{
+		testGCPCredential(t, &emailClaims{expectedEmails[0], true}, testAudience, testKeyID+"A", signerA),
+		testGCPCredential(t, &emailClaims{expectedEmails[1], true}, testAudience, testKeyID+"B", signerB),
+	}
+
+	// Returns a hardcoded JWK for token validation.
+	validatorClient := &http.Client{Transport: testJWKFetcher(t, jwks)}
+
+	// Validate.
+	emails, err := Validate(context.Background(), validatorClient, testTokens, testAudience)
+	if err != nil {
+		t.Fatalf("Validate error %v", err)
+	}
+
+	if !cmp.Equal(emails, expectedEmails) {
+		t.Errorf("Validate did not return expected emails: got %v, want %v", emails, expectedEmails)
+	}
+
+	// ValidateWithJWKS.
+	emails, err = ValidateWithJWKS(jwks, testTokens, testAudience)
+	if err != nil {
+		t.Fatalf("ValidateWithJWKS error %v", err)
+	}
+
+	if !cmp.Equal(emails, expectedEmails) {
+		t.Errorf("ValidateWithJWKS did not return expected emails: got %v, want %v", emails, expectedEmails)
+	}
+}
+
+func TestValidationError(t *testing.T) {
```

Comment:
nit: It's not clear from the test name or error msgs why the test should be a failure